oss-internship-2020/pffft/test_pffft_sandboxed.cc

```diff
@@ -0,0 +1,180 @@
+#include <assert.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/times.h>
+#include <syscall.h>
+#include <time.h>
+
+#include "fftpack.h"
+#include "pffft_sapi.sapi.h"
+#include "sandboxed_api/util/flag.h"
+#include "sandboxed_api/vars.h"
+
+ABSL_DECLARE_FLAG(string, sandbox2_danger_danger_permit_all);
+ABSL_DECLARE_FLAG(string, sandbox2_danger_danger_permit_all_and_log);
+
+class pffftSapiSandbox : public pffftSandbox {
+ public:
+  std::unique_ptr<sandbox2::Policy> ModifyPolicy(
+      sandbox2::PolicyBuilder*) override {
+    return sandbox2::PolicyBuilder()
+        .AllowStaticStartup()
+        .AllowOpen()
+        .AllowRead()
+        .AllowWrite()
+        .AllowSystemMalloc()
+        .AllowExit()
+        .AllowSyscalls({
+            __NR_futex,
+            __NR_close,
+            __NR_getrusage,
+        })
+        .DisableNamespaces()
+        .BuildOrDie();
+  }
+};
+
+double frand() { return rand() / (double)RAND_MAX; }
+
+double uclock_sec(void) { return (double)clock() / (double)CLOCKS_PER_SEC; }
+
+int array_output_format = 0;
+
+void show_output(const char* name, int N, int cplx, float flops, float t0,
+                 float t1, int max_iter) {
+  float mflops = flops / 1e6 / (t1 - t0 + 1e-16);
+  if (array_output_format) {
+    if (flops != -1) {
+      printf("|%9.0f   ", mflops);
+    } else
+      printf("|      n/a   ");
+  } else {
+    if (flops != -1) {
+      printf("N=%5d, %s %16s : %6.0f MFlops [t=%6.0f ns, %d runs]\n", N,
+             (cplx ? "CPLX" : "REAL"), name, mflops,
+             (t1 - t0) / 2 / max_iter * 1e9, max_iter);
+    }
+  }
+  fflush(stdout);
+}
+
+int main(int argc, char* argv[]) {
+  gflags::ParseCommandLineFlags(&argc, &argv, true);
+  int Nvalues[] = {64,    96,     128,        160,         192,     256,
+                   384,   5 * 96, 512,        5 * 128,     3 * 256, 800,
+                   1024,  2048,   2400,       4096,        8192,    9 * 1024,
+                   16384, 32768,  256 * 1024, 1024 * 1024, -1};
+  int i;
+
+  printf("initializing sandbox...\n");
```

Comment:
I suggest converting all such debug printf()'s in this code and elsewhere to something based on conditional logging, I believe glog can be used, where such messages are printed only when user starts this program with e.g. ./test_pffft_sandboxed -v=4 or somesuch.